server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java

```diff
@@ -533,15 +537,22 @@ private static Map<String, Mapper> buildMergedMappers(
             MergeReason reason,
             MapperMergeContext objectMergeContext
         ) {
-            Map<String, Mapper> mergedMappers = null;
+            Map<String, Mapper> mergedMappers = new HashMap<>(existing.mappers);
             for (Mapper mergeWithMapper : mergeWith) {
-                Mapper mergeIntoMapper = (mergedMappers == null ? existing.mappers : mergedMappers).get(mergeWithMapper.simpleName());
+                Mapper mergeIntoMapper = mergedMappers.get(mergeWithMapper.simpleName());
 
-                Mapper merged;
                 if (mergeIntoMapper == null) {
-                    merged = mergeWithMapper;
+                    if (mergeWithMapper instanceof ObjectMapper om) {
+                        ObjectMapper withoutMappers = om.withoutMappers();
```

Comment:
Could you help me understand the need for `withoutMappers` here please as well as the two branches of this conditional?

`mergeIntoMapper` is null, hence we are adding a new field, which I suspect is why we don't go through merge, we need to decrement the budget by the size of the current mapper don't we? But isn't the budget always one using `withoutMappers`? I see that we later use merge, which should take the limit into account as well, but then I don't understand the need for the initial `addFieldIfPossible` call before `merge`.

Reply:
I've added some comments to hopefully make it a little clearer.

Reply:
I think the new `MergeResult#addNewObjectMapper` method makes this a bit clearer, too.